File: src/main/java/dev/salmon/keystrokes/hud/api/KeyRenderer.java
package dev.salmon.keystrokes.hud.api;

import cc.polyfrost.oneconfig.platform.Platform;
import cc.polyfrost.oneconfig.renderer.NanoVGHelper;
import net.minecraft.client.gui.FontRenderer;
import net.minecraft.client.renderer.GlStateManager;

public final class KeyRenderer {
    private KeyRenderer() {}

    public static void drawBackground(float x, float y, float width, float height, KeyStyle style, boolean pressed, float fade, float scale) {
        int color = style.backgroundColor(pressed, fade);

        if (style.isRounded()) {
            float radius = style.getCornerRadius() * scale;
            NanoVGHelper.INSTANCE.setupAndDraw(true, vg ->
                    NanoVGHelper.INSTANCE.drawRoundedRect(vg, x, y, width, height, color, radius)
            );
        } else {
            Platform.getGLPlatform().drawRect(x, y, x + width, y + height, color);
        }
    }

    public static void drawCenteredText(FontRenderer fr, String text, float x, float y, float width, float height, KeyStyle style, boolean pressed, float fade) {
        int color = style.textColor(pressed, fade);
        boolean shadow = pressed ? style.isShadowActive() : style.isShadowIdle();

        float textX = x + (width - fr.getStringWidth(text) + 1) / 2f;
        float textY = y + (height - fr.FONT_HEIGHT + 2) / 2f;

        GlStateManager.enableBlend();
        fr.drawString(text, textX, textY, color, shadow);
        GlStateManager.disableBlend();
    }

    public static void drawSpaceGlyph(float x, float y, float width, float height, int color, boolean shadow) {
        float cx = x + width / 2f;
        float cy = y + height / 2f;

        drawHorizontalLine(cx - 6, cx + 5, cy - 0.5f, color);

        if (shadow) {
            if ((color & 0xFC000000) == 0) color |= 0xFF000000;
            int shadowColor = (color & 0xFCFCFC) >> 2 | (color & 0xFF000000);
            drawHorizontalLine(cx - 5, cx + 6, cy + 0.5f, shadowColor);
        }
    }

    private static void drawHorizontalLine(float startX, float endX, float y, int color) {
        if (endX < startX) {
            float tmp = startX; startX = endX; endX = tmp;
        }
        Platform.getGLPlatform().drawRect(startX, y, endX + 1, y + 1, color);
    }
}

File: src/main/kotlin/org/polyfrost/keystrokes/hud/api/KeyRenderer.kt
package org.polyfrost.keystrokes.hud.api

import cc.polyfrost.oneconfig.platform.Platform
import cc.polyfrost.oneconfig.renderer.NanoVGHelper
import net.minecraft.client.gui.FontRenderer
import net.minecraft.client.renderer.GlStateManager

object KeyRenderer {
    @JvmStatic
    fun drawBackground(
        x: Float,
        y: Float,
        width: Float,
        height: Float,
        style: KeyStyle,
        pressed: Boolean,
        fade: Float,
        scale: Float
    ) {
        val color = style.backgroundColor(pressed, fade)

        if (style.isRounded) {
            val radius = style.cornerRadius * scale
            NanoVGHelper.INSTANCE.setupAndDraw(
                true
            ) { vg: Long ->
                NanoVGHelper.INSTANCE.drawRoundedRect(
                    vg,
                    x,
                    y,
                    width,
                    height,
                    color,
                    radius
                )
            }
        } else {
            Platform.getGLPlatform().drawRect(x, y, x + width, y + height, color)
        }
    }

    @JvmStatic
    fun drawCenteredText(
        fr: FontRenderer,
        text: String?,
        x: Float,
        y: Float,
        width: Float,
        height: Float,
        style: KeyStyle,
        pressed: Boolean,
        fade: Float
    ) {
        val color = style.textColor(pressed, fade)
        val shadow = if (pressed) style.isShadowActive else style.isShadowIdle

        val textX = x + (width - fr.getStringWidth(text) + 1) / 2f
        val textY = y + (height - fr.FONT_HEIGHT + 2) / 2f

        GlStateManager.enableBlend()
        fr.drawString(text, textX, textY, color, shadow)
        GlStateManager.disableBlend()
    }

    @JvmStatic
    fun drawSpaceGlyph(x: Float, y: Float, width: Float, height: Float, color: Int, shadow: Boolean) {
        var color = color
        val cx = x + width / 2f
        val cy = y + height / 2f

        drawHorizontalLine(cx - 6, cx + 5, cy - 0.5f, color)

        if (shadow) {
            if ((color and -0x4000000) == 0) color = color or -0x1000000
            val shadowColor = (color and 0xFCFCFC) shr 2 or (color and -0x1000000)
            drawHorizontalLine(cx - 5, cx + 6, cy + 0.5f, shadowColor)
        }
    }

    private fun drawHorizontalLine(startX: Float, endX: Float, y: Float, color: Int) {
        var startX = startX
        var endX = endX
        if (endX < startX) {
            val tmp = startX
            startX = endX
            endX = tmp
        }
        Platform.getGLPlatform().drawRect(startX, y, endX + 1, y + 1, color)
    }
}